npm/pkg/controlplane/translation/parseSelector.go

```diff
@@ -238,3 +238,160 @@ func parseSelector(selector *metav1.LabelSelector) (labels []string, vals map[st
+type labelSelector struct {
+	// include is a flag to indicate whether Op exists or not.
+	include bool
+	settype ipsets.SetType
+	// setName is among
+	// 1. matchKey + ":" + matchVal (can be empty string) case
+	// 2. "matchKey" case
+	// or 3. "matchKey + : + multiple matchVals" case.
+	setName string
+	// members slice exists only if setType is only NestedLabelOfPod.
+	members []string
+}
+
+// parsedSelectors maintains slice of unique labelSelector.
+type parsedSelectors struct {
+	labelSelectors []labelSelector
+	// Use set data structure to avoid the duplicate setName among matchLabels and MatchExpression.
+	// The key of labelSet includes "!" if operator is "OpNOtIn" or "OpDoesNotExist"
+	// to make difference when it has the same key (and value), but different operator
+	// while this is weird since it is not always matched, but K8s accepts this spec.
+	labelSet map[string]struct{}
+}
+
+func newParsedSelectors() parsedSelectors {
+	return parsedSelectors{
+		labelSelectors: []labelSelector{},
+		labelSet:       map[string]struct{}{},
+	}
+}
+
+// addSelector only adds non-duplicated labelSelector.
+// Only nested labels from podSelector has members fields.
+func (ps *parsedSelectors) addSelector(include bool, setType ipsets.SetType, setName string, members ...string) {
+	setNameWithOp := setName
+	if !include {
+		setNameWithOp = "!" + setName
+	}
+
+	// in case setNameWithOp exists in a set, do not need to add it.
+	if _, exist := ps.labelSet[setNameWithOp]; exist {
+		return
+	}
+
+	ls := labelSelector{
+		include: include,
+		settype: setType,
+		setName: setName,
+		members: members,
+	}
+
+	ps.labelSelectors = append(ps.labelSelectors, ls)
+	ps.labelSet[setNameWithOp] = struct{}{}
+}
+
+// parseNSSelector parses namespaceSelector and returns slice of labelSelector object
+// which includes operator, setType, ipset name and always nil members slice.
+func parseNSSelector(selector *metav1.LabelSelector) []labelSelector {
+	// TODO(jungukcho): This will not happen
+	if selector == nil {
+		return []labelSelector{}
+	}
+
+	parsedSelectors := newParsedSelectors()
+
+	// #1. All namespaces case
+	if len(selector.MatchLabels) == 0 && len(selector.MatchExpressions) == 0 {
+		parsedSelectors.addSelector(true, ipsets.KeyLabelOfNamespace, util.KubeAllNamespacesFlag)
+		return parsedSelectors.labelSelectors
+	}
+
+	// #2. MatchLabels
+	for matchKey, matchVal := range selector.MatchLabels {
+		// matchKey + ":" + matchVal (can be empty string) case
+		setName := matchKey + util.IpsetLabelDelimter + matchVal
+		parsedSelectors.addSelector(true, ipsets.KeyValueLabelOfNamespace, setName)
+	}
+
+	// #3. MatchExpressions
+	for _, req := range selector.MatchExpressions {
+		var setName string
+		var setType ipsets.SetType
+		switch op := req.Operator; op {
+		case metav1.LabelSelectorOpIn, metav1.LabelSelectorOpNotIn:
+			// "(!) + matchKey + : + matchVal" case
+			setName = req.Key + util.IpsetLabelDelimter + req.Values[0]
```

Comment:
can you use the util.GetIpSetFromLabelKV func

Reply:
Can you add a comment here/above this function on why we will never call parseNSSelector with a matchExpression with multiple values, we should be flattening it out before calling this function.